D3ploy update

FlanFlanagan opened the pull request:
@jbae11 @gwenchee 

First cut at updating cycamore for d3ploy.

scopatz commented on src/fuel_fab.cc:
```diff
@@ -507,6 +507,13 @@ void FuelFab::RecordPosition() {
+void FuelFab::Tick() {
+  cyclus::toolkit::RecordTimeSeries<double>(fuelfab_record_input, this, 
+                                            throughput);
+  cyclus::toolkit::RecordTimeSeries<double>(fuelfab_record_output, this, 
```
hmm shouldn't these be the input and output commodity names?

FlanFlanagan replied:
D3ploy needs unique calls for these.  A fuelfab and a reactor might share the same commodity to trade fuel, but in order to compare the supply and demand of that commodity the facilities need their own unique tags for that commodity.

scopatz commented on src/reactor.h:
```diff
@@ -415,6 +415,15 @@ class Reactor : public cyclus::Facility,
 
+  // For new Record Time Series functionality
+  #pragma cyclus var { \
+    "default": "RxFuelDemand", \
```
Should this be something like FreshFuelDemand?

scopatz replied:
Or ReactorFuelDemand?  I don't think defaults should contain abbreviations. It is harder for new folks to understand

scopatz commented on src/separations.h:
```diff
@@ -206,6 +206,20 @@ class Separations
   std::map<std::string, cyclus::toolkit::ResBuf<cyclus::Material> > streambufs;
 
+  #pragma cyclus var { \
+    "default": "separations_record_input", \
```
Its seems like there should be a naming convention rather than switching between `lower_with_underscores` and `CapCase`

scopatz commented:
Thanks for putting this in @FlanFlanagan! Looks like there is a bug with FuelFab::Tick()

jbae11 commented:
Is there a reason why some are called `xxx_record_input` and some are called `xxx_record_incommod`?

FlanFlanagan commented:
Just my own inconsistencies, I'll look into it. Thanks!

jbae11 commented on src/mixer.cc:
```diff
@@ -96,6 +96,10 @@ void Mixer::EnterNotify() {
 
 void Mixer::Tick() {
+  cyclus::toolkit::RecordTimeSeries<double>(mixer_record_demand, this, 
+                                            throughput);
```
Shouldn't this be `std::min(throughput / mixing_ratios[indx], in_buf_sizes[indx])` for each stream?

FlanFlanagan replied:
At the moment we do not support vectors in the RecordTimeSeries, it's fixed to doubles or strings at the moment I believe. Basically RecordTimeSeries uses the first parameter to indicate what to name the column in the cyclus output database. Making it a vector would require a rewrite of the output database I think?

You're welcome to do that if you'd like 😛 .

FlanFlanagan replied:
Or just a function in the cyclus backend that take a RecordTimeSeries(vector) and turn it into a bunch of individual ones.

FlanFlanagan replied:
The other option would be to make a for loop in the facility that loops through individual isotopes and records those.

jbae11 commented on src/fuel_fab.cc:
```diff
@@ -507,6 +507,13 @@ void FuelFab::RecordPosition() {
 
+void FuelFab::Tick() {
+  cyclus::toolkit::RecordTimeSeries<double>(fuelfab_record_demand, this, 
+                                            throughput);
```
Instead of throughput, wouldn't this be:
1. `fill_size - fill.quantity()`
2. `fiss_size - fiss.quantity()`
Since that is the remaining capacity for accepting?

jbae11 replied:
yup. I think the recordtimeseries would be best if it was a vector.

FlanFlanagan commented on src/separations.cc:
```diff
@@ -103,6 +103,10 @@ void Separations::EnterNotify() {
+  cyclus::toolkit::RecordTimeSeries<double>(separations_record_demand, this, 
+                                            throughput);
+  cyclus::toolkit::RecordTimeSeries<double>(separations_record_supply, this, 
+                                            throughput);
```
I'm not sure it does. We aren't tracking what it actually did that time step, but rather what it is capable of that time step.

jbae11 replied:
yeah but processing same amounts of spent fuel may yield different output product quantity depending on the spent fuel composition..right? For example reprocessing 1kg of UNF with 10% plutonium  and 15% plutonium would allow it to output different quantities of pu?

FlanFlanagan replied:
Oh yes, true. The supply could be what it actually produces that time step. I still think demand should be based on it's throughput or remaining inventory size.

jbae11 replied:
yup! I agree on the demand

FlanFlanagan commented on src/storage.cc:
```diff
@@ -117,6 +117,11 @@ void Storage::Tick() {
+  cyclus::toolkit::RecordTimeSeries<double>(storage_record_demand, this, 
+                                            throughput);
+  cyclus::toolkit::RecordTimeSeries<double>(storage_record_supply, this, 
+                                            throughput);
```
That might be correct. I suppose it should actually be max(throughtput, *remaining inventory size*). Also we don't care about what it actually does, but rather what it is capable of for this facility.

jbae11 replied:
Hmm, but wouldn't the maximum it can output per timestep be `min(throughput, remaining inventory)`?

FlanFlanagan replied:
Oh yes, min.

FlanFlanagan commented on src/reactor.h:
```diff
@@ -415,6 +415,15 @@ class Reactor : public cyclus::Facility,
+    "uilabel": "Demand Tracking String", \
+    "doc": "String used to record the amount of fuel demanded "\
+            "by the reactor at a given timestep" \
+  }
```
That's a good point. I can add that. I even mentioned it at a conference recently as a driver for deployment of repositories.

FlanFlanagan commented on src/reactor.cc:
```diff
@@ -407,6 +407,8 @@ void Reactor::Load() {
   ss << n << " assemblies";
   Record("LOAD", ss.str());
+  cyclus::toolkit::RecordTimeSeries<double>(reactor_fuel_demand, this, 
+                                         n*n_assem_batch*assem_size);
```
That's why it's n for the number of batches. The number of batches (n) varies to account for this.

jbae11 replied:
I'm not sure if n is the number of batches. Maybe I'm reading something wrong.
Wouldn't `n` be the `number of assemblies per batch` for usual cases?
```
 int n = std::min(n_assem_core - core.count(), fresh.count());
```

FlanFlanagan replied:
'''  
#pragma cyclus var { \
    "uilabel": "Number of Assemblies per Batch", \
    "doc": "Number of assemblies that constitute a single batch.  " \
           "This is the number of assemblies discharged from the core fully " \
           "burned each cycle."           \
           "Batch size is equivalent to ``n_assem_batch / n_assem_core``.", \
  }
  int n_assem_batch;
  #pragma cyclus var { \
    "default": 3, \
    "uilabel": "Number of Assemblies in Core", \
    "uitype": "range", \
    "range": [1,3], \
    "doc": "Number of assemblies that constitute a full core.", \
  }
  int n_assem_core
'''

FlanFlanagan replied:
Ah so what I actually want to do is '''n*assem_size'''. Which should be number of assemblies * mass per assembly? Not sure why I put the number of batches in there. Good catch thanks!

FlanFlanagan commented:
I have updated this PR. It now should work with the new version of d3ploy (still in PR). Only the reactor / source / sink have been updated. I removed the functionality from the rest to think about those further. 

Due to the d3ploy update you can no longer easily reframe the supply - demand problem (spent used fuel being viewed as a demand for storage space, for example). To account for this I hard coded that behavior into the reactor and sink.

gwenchee commented:
I tried installing the new Cycamore updates, and encountered this error: 

```
[100%] Linking CXX executable bin/cycamore_unit_tests
lib/cyclus/cycamore/libcycamore.so: undefined reference to `VTT for cycamore::FuelFab'
lib/cyclus/cycamore/libcycamore.so: undefined reference to `cycamore::Reactor::RecordSideProduct(bool)'
lib/cyclus/cycamore/libcycamore.so: undefined reference to `vtable for cycamore::FuelFab'
collect2: error: ld returned 1 exit status
CMakeFiles/cycamore_unit_tests.dir/build.make:485: recipe for target 'bin/cycamore_unit_tests' failed
make[2]: *** [bin/cycamore_unit_tests] Error 1
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/cycamore_unit_tests.dir/all' failed
make[1]: *** [CMakeFiles/cycamore_unit_tests.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2
Traceback (most recent call last):
  File "install.py", line 214, in <module>
    main()
  File "install.py", line 211, in main
    install_cyclus(args)
  File "install.py", line 107, in install_cyclus
    shell=(os.name == 'nt'))
  File "/home/gwenchee/anaconda3/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make']' returned non-zero exit status 2.
```

The command i put into the cli is: 
``` 
python install.py --clean-build
```

FlanFlanagan commented:
Yes, I'm working on figuring this out right now.
